HDFOutputer.cc

```diff
@@ -0,0 +1,163 @@
+#include "HDFOutputer.h"
+#include "summarize_serializers.h"
+#include "lz4.h"
+#include <iostream>
+#include <cstring>
+#include <cmath>
+#include <set>
+
+constexpr int max_batch_size = 2; 
+
+using namespace cce::tf;
+using product_t = std::vector<char>; 
+
+HDFOutputer::HDFOutputer(std::string const& iFileName, unsigned int iNLanes) : 
+  file_(hdf5::File::create(iFileName.c_str())),
+  serializers_{std::size_t(iNLanes)},
+  serialTime_{std::chrono::microseconds::zero()},
+  parallelTime_{0}
+  {}
+
+HDFOutputer::~HDFOutputer() { }
+
+void HDFOutputer::setupForLane(unsigned int iLaneIndex, std::vector<DataProductRetriever> const& iDPs) {
+  auto& s = serializers_[iLaneIndex];
+  s.reserve(iDPs.size());
+  dataProductIndices_.reserve(iDPs.size());
+  for(auto const& dp: iDPs) {
+    s.emplace_back(dp.name(), dp.classType());
+  }
+   products_.reserve(10000);
+   events_.reserve(100);
+   offsets_.reserve(10000);
+}
+
+void HDFOutputer::productReadyAsync(unsigned int iLaneIndex, DataProductRetriever const& iDataProduct, TaskHolder iCallback) const {
+  auto& laneSerializers = serializers_[iLaneIndex];
+  auto group = iCallback.group();
+  laneSerializers[iDataProduct.index()].doWorkAsync(*group, iDataProduct.address(), std::move(iCallback));
+}
+
+void HDFOutputer::outputAsync(unsigned int iLaneIndex, EventIdentifier const& iEventID, TaskHolder iCallback) const {
+  auto start = std::chrono::high_resolution_clock::now();
+  queue_.push(*iCallback.group(), [this, iEventID, iLaneIndex, callback=std::move(iCallback)]() mutable {
+      auto start = std::chrono::high_resolution_clock::now();
+      const_cast<HDFOutputer*>(this)->output(iEventID, serializers_[iLaneIndex]);
+        serialTime_ += std::chrono::duration_cast<decltype(serialTime_)>(std::chrono::high_resolution_clock::now() - start);
+      callback.doneWaiting();
+    });
+    auto time = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - start);
+    parallelTime_ += time.count();
+}
+
+void HDFOutputer::printSummary() const  {
+  std::cout <<"HDFOutputer\n  total serial time at end event: "<<serialTime_.count()<<"us\n"
+    "  total parallel time at end event: "<<parallelTime_.load()<<"us\n";
+  summarize_serializers(serializers_);
+}
+
+std::pair<product_t, std::vector<size_t>> 
+HDFOutputer::
+get_prods_and_sizes(std::vector<product_t> & input, 
+         int prod_index, 
+         int stride) {
+  product_t products; 
+  std::vector<size_t> sizes;
+  sizes.reserve(input.size()); 
+  // or may use (std::ceil(double(input.size()-prod_index)/stride)); 
+  for(int j = prod_index; j< input.size(); j+=stride) {
+    sizes.push_back(offsets_[prod_index]+=input[j].size());
+    products.insert(end(products), std::make_move_iterator(begin(input[j])), std::make_move_iterator(end(input[j])));
+  }
+  return {products, sizes};
+}
+
+template <typename T> 
```

Comment:
How about placing the next two functions in an anonymous namespace? They don't need to be in the global namespace.

Reply:
Just the `write_ds` function? all the others are class members.

Reply:
Sorry, I mis-read the code a bit so only `write_ds`.